Fixes to seeking and auto restore

SuuperW opened the pull request:
The current implementation of TAStudio is a gigantic mess and has various pesky bugs. This is an attempt to resolve some of these issues.

Changes that are not bug fixes:
1) Show seek progress bar even for short seeks.
2) Include hotkey to go to green arrow, default is `R`.
3) Middle-click resumes a seek if a seek was in progress. (was this a bug? not sure)
4) Changed Lua API method `tastudio.setplayback` to return as soon as the seek begins. This is basically a side-effect of fixing bugs. See comments below for reasoning. (https://github.com/TASEmulators/BizHawk/pull/4347#issuecomment-2964930812)
5) Removed Lua API method `client.seekframe`. It would freeze the user interface for the duration of the seek, same as setplayback.

List of bugs fixed (also in commit message):
1) Auto-restore did not work when painting "axis" inputs.
2) If unpaused in recording mode, manual seek to frame A (click cursor column) then before that seek finishes seek to frame B. It would not unpause after reaching frame B to resume recording.
3) TAStudio would fail to pause with auto-restore if a second edit was made to a non-greenzoned frame while auto-restore seek was in progress.
4) Canceling seek would not remove the seek progress bar.
5) Mouse drag seeking was broken while seeking. (It would not change the target seek frame until prior seek had completed.)
6) Multiple issues with seeking to frames past the current movie length.
7) Modifier key + right click would jump to the last edited frame (even if this right click made no edits).
8) Clicking on an axis value while in axis editing mode would disable recording mode, regardless of whether an edit was made.
9) Clicking on an axis value while in axis editing mode would trigger auto-restore, even if no edit was made (seek to last edit frame).
EDIT: I'm not listing them all here anymore, see commit comments for more details. But, somewhere along the line I accidentally fixed these:
1) It used to be possible to invalidate greenzone without rewinding by pressing escape after mouse editing an axis value.
2) Right-click + scroll up (to seek backwards) immediately after opening TAStudio used to break the emulator pause functionality; it would be stuck unpaused.

I've done my best to avoid changing any behavior that isn't a bug. If you see any, please let me know.

Check if completed:
- [x] I have run any relevant test suites
- [x] I, the commit author, have read the [licensing terms for contributors](https://github.com/TASEmulators/BizHawk/blob/master/contributing.md#copyrights-and-licensing) (last updated 2024-06-22) and am compliant

SuuperW committed:
move safety check inside StartSeeking, instead of making caller responsible

SuuperW committed:
rename variable

SuuperW committed:
tastudio should own tastudio logic

SuuperW committed:
Make TAStudio take ownership of its own seeking logic and simplify some code, fixing a few bugs:
1) Auto-restore did not work when painting "axis" inputs.
2) If unpaused in recording mode, manual seek to frame A (click cursor column) then before that seek finishes seek to frame B. It would not unpause after reaching frame B to resume recording.
3) TAStudio would fail to pause with auto-restore if a second edit was made to a non-greenzoned frame while auto-restore seek was in progress.
4) Canceling seek would not remove the seek progress bar.

SuuperW committed:
Remove obsolete conditional.

SuuperW committed:
Fix: Mouse drag seeking was broken while seeking. (It would not change the target seek frame until prior seek had completed.)

SuuperW committed:
GoToFrame's "future frame" cases were just buggy versions of the normal case. So just use the code of StartAtNearestFrameAndEmulate instead. Fixes multiple bugs when seeking to frames past the end of the movie.

SuuperW committed:
Fix: modifier key + right click would jump to the last edited frame (even if this right click made no edits)

SuuperW committed:
simplify a bit of logic

SuuperW committed:
Fix: Clicking on an axis value while in axis editing mode would disable recording mode, regardless of whether an edit was made.
Fix: Clicking on an axis value while in axis editing mode would trigger auto-restore, even if no edit was made (seek to last edit frame).

SuuperW commented:
added more fixes
plenty more still to come

SuuperW committed:
Windows weirdness: MouseUp event is not guaranteed to be raised.

SuuperW committed:
Replace some auto restore code with something simpler. Fixes auto-restore for right-click edits.

SuuperW committed:
Revert commit a1c8c4ed3a676954358cb493efc878436bd43a4e. It was always pointless because it didn't fix the issue it was supposed to fix, and the later commit that actually fixes it was all that was needed all along.

SuuperW committed:
Convert remaining auto-restore points to the new system. Also include refreshing in FrameEdited. Fixes more bugs.

SuuperW committed:
inline function that is only called from one place

SuuperW committed:
Fix: turbo seek did not work when navigating while unpaused

vadosnaprimer commented:
What matters is to mention them in commits, otherwise it'd be a pain to review.

Morilli commented:
> 2. If unpaused in recording mode, manual seek to frame A (click cursor column) then before that seek finishes seek to frame B. It would not unpause after reaching frame B to resume recording.

Does that even make sense though? Maybe seeking while unpaused *and* in recording mode actually *should* pause once seeking finishes? I generally do not really see a situation in which you would want to seek somewhere while unpaused and in recording mode, that sounds like a recipe to lose inputs.

SuuperW committed:
Call our new FrameEdited thing for auto restore whenever greenzone is invalidated, instead of manually calling at each edit point. This fixes auto-restore for undo/redo actions.

SuuperW committed:
Fix: Greenzone would be invalidated if pasting or clearing a frame even if no change was made.

SuuperW committed:
Fix: Seeks from Lua would freeze the UI and not return to Lua until the seek has completed, making the Lua script unable to see the frames during the seek.

vadosnaprimer commented:
> Most recent commit (fix for Lua) changes the behavior of the Lua API. Previously, `tastudio.setplayback` would not return until seeking had completed and no Lua events/callbacks would happen during the seek. Now it returns when seeking begins (after load state, if that happened).

Sounds kinda weird. Callbacks being ignore while it's seeking may or may not be desirable (I never used lua to control tastudio/taseditor behavior in real tasing), but "set playback" implies a concrete event which is defined in that function's description:

> Seeks the given frame (a number) or marker (a string)

The goal is getting to that frame before we do something. After it finishes, it is expected that it's now safe to do whatever we wanted to do on that frame (or marker). Especially if there has to be a lot of seeking due to lack of states on those frames: if we say "Hey tastudio, please put playback on frame 9000!" we won't be satisfied if it puts it on frame 1000 and says "Done!!!".

> the old behavior is very bad

Why?

SuuperW commented:
> The goal is getting to that frame before we do something. After it finishes, it is expected that it's now safe to do whatever we wanted to do on that frame (or marker). Especially if there has to be a lot of seeking due to lack of states on those frames: if we say "Hey tastudio, please put playback on frame 9000!" we won't be satisfied if it puts it on frame 1000 and says "Done!!!".

This is at least trivial to fix in Lua: `while emu.framecount() < targetFrame do emu.frameadvance() end`

> Why?

1) It freezes the user interface. If TAStudio has to seek 8000 frames and the user can only emulate 200fps, it will be 40 seconds until ANYTHING is displayed and the user can take any further action. This would appear to be BizHawk crashing. This also means that it is impossible to cancel the seek, for example if the user didn't realize they were about to re-emulate thousands of frames.
2) During this time no Lua events can run. This means that any Lua functionality that depends on monitoring playback or keeping track of values over time can never work in a script that uses `tastudio.setplayback`. For example a script that I regularly use requires data from the previous frame (sometimes two) in order to display correct information on the current frame, and that is impossible if I use `tastudio.setplayback(frame_i_want_to_go_to)`.

SuuperW commented:
The freeze is forced no matter what.

> This sounds like you have turbo seeking enabled and scripts during turbo disabled. If that's not the case, then I agree that it's also a problem.
It freezes without turbo seek. I rarely have turbo seek enabled, and never turn off scripts during turbo.

I don't know what we could do to allow the script to run during the seek, without returning from `tastudio.setplayback`. It would not be too hard if the script uses the after-frame event, but since the "default" way to run some Lua code every frame is to run an infinite loop with `emu.frameadvance()` in it, ... well, I don't know how to make that work. The script will obviously be at the call to `tastudio.setplayback`, and I don't think there's any sensible way to move execution to wherever in the script `emu.frameadvance()` is.

vadosnaprimer commented:
Maybe I'm seeing it now. So what actually happens when I do `setplayback(9000)` from frame 10, given the most recent changes? Will it return from that function but actually keep seeking to 9000 while running all the tools normally?

What about `client.seekframe()`?

SuuperW commented:
Also, if the UI did not freeze then the user would be able to pause emulation and interrupt the seek. This will be much easier to handle if `tastudio.setplayback` has already returned. If it hasn't, and Lua is still waiting for the seek to complete, what happens if the seek never completes? We cannot possibly guarantee that emulation will ever reach the frame specified in the call to `tastudio.setplayback`, and suspending the Lua indefinitely would be very confusing behavior ... as would *sometimes* returning on the wrong frame but not always.

SuuperW commented:
> Maybe I'm seeing it now. So what actually happens when I do `setplayback(9000)` from frame 10, given the most recent changes? Will it return from that function but actually keep seeking to 9000 while running all the tools normally?
> 
> What about `client.seekframe()`?

There are two possibilities here.
First, if the movie does not have 9000 frames, setplayback will return immediately, doing nothing. (I do not know why this check exists.)
If the movie does have 9000 frames, the new setplayback behavior is to return right after the seek begins. It will keep seeking while all the tools run normally including the Lua script, which must explicitly check the current frame to know when frame 9000 is reached.

In either situation, seekframe will freeze the UI and seek, the same as old setplayback behavior (for when the movie did have 9000 frames) and just as it would in the current master branch since I haven't modified that code

vadosnaprimer commented:
Yeah okay. As long as it keeps seeking (until you disable it) returning right away is better.

SuuperW committed:
add hotkey for go to green arrow (manual restore position)
Also, fix: Middle-click restore would not update seek begin frame, potentially causing seek progress indicator to be wrong.

SuuperW committed:
Feature, from TASeditor: middle-click resumes seek if seek is in progress. This might be what StartSeeking's fromMiddleClick parameter (removed in last commit) was attempting to do.

SuuperW committed:
TAStudio should own TAStudio logic. Fix: The green arrow would move when making multiple edits with auto-restore off if the edit caused a seek of >1 frame.

SuuperW committed:
Deprecate client.seekframe. This had the same issue as tastudio.setplayback, freezing the UI. Since seekframe cannot go backwards, updating it to return immediately results in it doing absolutely nothing.
Note that it never was doing a "seek" as defined by MainForm, so we aren't removing that feature. And turbo-seek isn't relevant both because it wasn't a seek and because currently the only way to have a turbo-seek is to use the Play Movie dialog. If true seeking is desired a new lua method should be made. Also also, it did not actually touch InvisibleEmulation.

SuuperW committed:
Fix: Manual seeks past the end of the movie were not respected if AutoPause was on.

SuuperW committed:
Show seek progress bar if we are seeking more than 1 frame. We don't know how long it's going to take! Also if the user ends up pausing there should be a visual indication of seeking.

SuuperW committed:
Remove redundant code.

SuuperW commented:
My list of things to change in this branch is now empty. Please review.

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -773,29 +764,114 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+
+			bool ret = false;
+			if (_batchEditMinFrame != -1)
+				ret = FrameEdited(_batchEditMinFrame);
```
Missing braces in this and the following method

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -773,29 +764,114 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+		/// <summary>
+		/// Begins a batch of edits, for auto-restore purposes. Auto-restore will be delayed until EndBatchEdit is called.
+		/// </summary>
+		public void BeginBatchEdit(bool fromLua = false)
```
What happens if there are mismatched begin and end calls?

SuuperW replied:
An extra call to end would do nothing. A second call to begin before calling end would also do nothing. A call to begin without a call to end would result in not rewinding/auto-restoring at the right time, possibly emulating outside the greenzone (since no rewind), and maybe refreshes not happening at the right times.

Nested calls would just be an end at the first call to end, all future edits would be non-batched (no harm except doing needless work including repeated refreshing... which actually can be a bit of harm if there are tons of edits, for example in current master doing "Clone # Times" and entering something in the thousands will refresh that many times which takes a while during which the UI is unresponsive).

If you are worried that future edits might accidentally result in missing a `EndBatchEdit` call, there's already an unconditional call in list view MouseUp, so there's that at least. All calls currently should be matched except the one in list view MouseUp because it doesn't need a begin call as-is.

In current master it is possible to avoid the MouseUp event handler in Windows by pressing Alt+Tab to switch to another window while the mouse button is held. This is fixed in this branch by calling MouseUp when the window is "deactivated" (read: loses focus). If you know of any other ways to skip the MouseUp handler let me know how.

They are public because they are called from the Lua API, via `tastudio.applyinputchanges`.

YoshiRulz replied:
So it won't be put in an unexpected state, and any weird behaviour would only last until you click again? That's good enough for me.

SuuperW commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.cs:
```diff
@@ -549,7 +553,7 @@ private bool StartNewMovieWrapper(ITasMovie movie, bool isNew)
 			movie.InputRollSettingsForSave = () => TasView.UserSettingsSerialized();
 			movie.BindMarkersToInput = Settings.BindMarkersToInput;
-			movie.GreenzoneInvalidated = GreenzoneInvalidated;
+			movie.GreenzoneInvalidated = (f) => FrameEdited(f);
```
OK.  ...but why? Is this one of those silly style guidelines? `_ = FrameEdited(f)` looks much uglier to me.
I'd have done the latter but that would require changing the type of GreenzoneInvalidated which seems a bit silly since the return value won't be used there either.

YoshiRulz replied:
Uglier, sure, but it communicates that the LHS('s caller) doesn't use the implicit lambda return.
And yeah it doesn't look like changing it would be useful.

SuuperW committed:
Fix: tastudio.setplayback would refuse to seek past the end of a movie

SuuperW committed:
Fix: Using tastudio.setplayback to seek to a non-existent marker would permanently suppress Lua.
Fix: Using tastudio.setplayback with a Lua number that happens to not currently be represented as an integer would throw.

SuuperW commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.Navigation.cs:
```diff
@@ -5,73 +5,44 @@ namespace BizHawk.Client.EmuHawk
+		public void GoToFrame(int frame, bool OnLeftMouseDown = false)
 		{
-			if (frame != Emulator.Frame) // Don't go to a frame if you are already on it!
+			if (frame == Emulator.Frame)
```
See my branch SuuperW:test and see if you like it.

YoshiRulz replied:
Ehh it's marginally better. You can leave it like this or try moving this method after `GoToPreviousMarker`, idrc.

YoshiRulz replied:
Ehh it's marginally better. You can leave it like this or try moving this method after `GoToPreviousMarker`, idrc.

YoshiRulz commented on src/BizHawk.Client.Common/movie/tasproj/TasMovie.Editing.cs:
```diff
@@ -82,13 +83,17 @@ public void SetFrame(int frame, string source)
 		{
-			ChangeLog.AddGeneralUndo(frame, frame, $"Clear Frame: {frame}");
+			string empty = Bk2LogEntryGenerator.EmptyEntry(Session.MovieController);
+			if (GetInputLogEntry(frame) != empty)
```
`if (GetInputLogEntry(frame) == empty) return;`

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/Lua/Libraries/TAStudioLuaLibrary.cs:
```diff
@@ -143,30 +143,32 @@ public void SetPlayback(object frame)
-					f = Tastudio.CurrentTasMovie.Markers.FindIndex((string)frame);
-					if (f == -1)
+					f = Tastudio.CurrentTasMovie.Markers.FindIndex((string) frame);
+					if (f != -1)
```
`>= 0`, and I think this should be using a separate temp variable rather than reassigning `f`.

SuuperW replied:
I disagree with `>= 0`. `FindIndex` specifies that it returns -1 if no match is found. Checking for that specific value makes it clear what the code intends to do. And if we don't want to trust `FindIndex` to work correctly then we should be checking for the upper bound of valid indexes too.

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.IToolForm.cs:
```diff
@@ -94,18 +94,25 @@ protected override void UpdateAfter()
 			}
 
+			if (!_seekingByEdit)
+				_shouldMoveGreenArrow = true;
```
missed one

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -523,10 +518,16 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+					}
+
+					var record = CurrentTasMovie[RestorePositionFrame];
+					if (!record.Lagged.HasValue)
```
`record.Lagged is null`

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -773,29 +763,127 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+
+		public void BatchEdit(Action action)
+		{
+			// This is only caled from Lua.
```
If it's only called from Lua, put "Lua" or "ApiHawk" in the method name.
(Per the other thread, I think it's fine to keep the separate begin/end calls everywhere else.)

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -773,29 +763,127 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+			GreenzoneInvalidatedCallback?.Invoke(frame); // lua callback
+
+			if (CurrentTasMovie.LastEditWasRecording)
+				return false;
```
missed one

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -773,29 +763,127 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+
+			if (needsRefresh)
+			{
+				if (TasView.IsPartiallyVisible(frame) || frame < TasView.FirstVisibleRow)
```
May be faster?
```suggestion
				if (frame <= TasView.LastVisibleRow || TasView.IsPartiallyVisible(frame))
```

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -773,29 +763,127 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+				else if (TasView.RowCount != CurrentTasMovie.FrameCount + 1)
+				{
+					// Row count must always be kept up to date even if last row is not directly visible.
+					TasView.RowCount = CurrentTasMovie.InputLogLength + 1;
```
Is this meant to be a different prop to the one in the condition?

SuuperW replied:
No, I didn't even realize there were two such properties. What is the theoretical difference? It's not explained in the property comments and their implementations are identical.

YoshiRulz commented:
To fix the label in the Hotkeys dialog, increase this to 120 dp:
https://github.com/TASEmulators/BizHawk/blob/a030b16916015a86fe0b12028c67f8f07460673f/src/BizHawk.Client.EmuHawk/config/HotkeyConfig.cs#L98

SuuperW commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -1388,17 +1435,10 @@ public void EditAnalogProgrammatically(KeyEventArgs e)
 
-			RefreshDialog();
+			if (!EndBatchEdit() && (prevTyped != _axisTypedValue || !AxisEditingMode))
+				RefreshDialog();
```
silly me

YoshiRulz commented on src/BizHawk.Client.EmuHawk/config/HotkeyConfig.Designer.cs:
```diff
@@ -198,7 +198,7 @@ private void InitializeComponent()
 			this.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font;
 			this.CancelButton = this.IDB_CANCEL;
-			this.ClientSize = new System.Drawing.Size(753, 463);
+			this.ClientSize = new System.Drawing.Size(783, 463);
```
Are any of these changes necessary? Just changing `iwOffsetX` was sufficient for me with Mono.

SuuperW replied:
The window was not wide enough to see all the controls on the General tab when I only changed `iwOffsetX`. The button for "special tricky bindings" was cut off on the third column.

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -773,29 +763,129 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+			_batchEditing = false;
+			if (_batchEditMinFrame != -1)
+			{
+				return FrameEdited(_batchEditMinFrame);
```
I know I said you shouldn't worry about mismatched calls, but `_batchEditMinFrame = -1;` here could prevent some bugs.

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -773,29 +763,129 @@ private void TasView_MouseDown(object sender, MouseEventArgs e)
+			// This is only caled from Lua.
+			_editIsFromLua = true;
+			BeginBatchEdit();
+			action();
```
...and `try`/`finally` here

SuuperW replied:
The Lua call shouldn't be able to throw, but I suppose this could be good if custom tools use it?

SuuperW committed:
Safety for any custom or future tools that want to use this.

YoshiRulz commented:
Is bug fix 2) in OP the same as #2329?

SuuperW commented:
Probably not, since he doesn't mention recording mode. It's more likely he was experiencing bug 3, but none of his posts indicate the steps taken to reproduce the bug so I can't tell.

Morilli commented on src/BizHawk.Client.EmuHawk/tools/TAStudio/TAStudio.ListView.cs:
```diff
@@ -69,38 +68,21 @@ private int AxisEditRow
-			WasRecording = CurrentTasMovie.IsRecording() || WasRecording;
-			TastudioPlayMode(); // suspend rec mode until seek ends, to allow mouse editing
+			_seekingTo = frame;
+			MainForm.PauseOnFrame = int.MaxValue; // This being set is how MainForm knows we are seeking, and controls TurboSeek.
```
Is there any reason you changed this to be `int.MaxValue` instead of `frame`? This breaks numerous (reasonable) assumptions in the code that `PauseOnFrame` actually contains the correct frame number value and leads to incorrect behavior with turbo seeking because the frame advance code does not know when the turbo seek ends anymore.

SuuperW replied:
What specific incorrect behavior are you seeing? The only reference to `PauseOnFrame` that care about the value other than checking if it is `null` or `<= Emulator.Frame` is MainForm line 3018. That might be an oversight by me (looks like it might cause the core to not render the last frame, for cores that support skipping rendering). But it's not related to frame advance.

For everything else, TAStudio handles things by setting `PauseOnFrame` to `null` at the correct time.

I would guess that the reason I set it to `MaxValue` was to avoid potential issues with having it get out of sync with `_seekingTo`, e.g. when using mouse wheel to change the seek frame. But the correct solution to that is probably to ~~change `_seekingTo` to a property and use `MainForm.PauseOnFrame` as a backing "field"~~. EDIT: Actually that would make it so that MainForm's pause-after-seek logic applies for turbo seeks, which is different from TAStudio's.

Morilli replied:
Well it kinda breaks the tooltips in `SetPauseStatusBarIcon` although that's a really minor issue.

But yes, it also directly regresses #2081. Additionally this will now call `UpdateAfter` twice for lua scripts, once in `StepRunLoop_Core` because `atTurboSeekEnd` isn't set, and once in the `StopSeeking` call, which causes everything to be rendered twice which is very noticable with transparent drawings.

SuuperW replied:
See #4743
